Source/Modules/php.cxx

```diff
@@ -372,8 +372,13 @@ class PHP : public Language {
     Printf(s_header, "int error_code;\n");
     Printf(s_header, "ZEND_END_MODULE_GLOBALS(%s)\n", module);
     Printf(s_header, "ZEND_DECLARE_MODULE_GLOBALS(%s)\n", module);
-    Printf(s_header, "#define SWIG_ErrorMsg() ZEND_MODULE_GLOBALS_ACCESSOR(%s, error_msg)\n", module);
-    Printf(s_header, "#define SWIG_ErrorCode() ZEND_MODULE_GLOBALS_ACCESSOR(%s, error_code)\n", module);
+    Printf(s_header, "#ifdef ZTS\n");
+    Printf(s_header, "#define SWIG_ErrorMsg() TSRMG(%s_globals_id, zend_%s_globals *, error_msg )\n", module, module);
+    Printf(s_header, "#define SWIG_ErrorCode() TSRMG(%s_globals_id, zend_%s_globals *, error_code )\n", module, module);
+    Printf(s_header, "#else\n");
+    Printf(s_header, "#define SWIG_ErrorMsg() (%s_globals.error_msg)\n", module);
+    Printf(s_header, "#define SWIG_ErrorCode() (%s_globals.error_code)\n", module);
+    Printf(s_header, "#endif\n\n");
```

Comment:
I don't understand why this change is necessary - looking at the PHP 7.1 headers , when `ZTS` is defined `Zend/zend_API.h` defines `ZEND_MODULE_GLOBALS_ACCESSOR` as:

```
#define ZEND_MODULE_GLOBALS_ACCESSOR(module_name, v) ZEND_TSRMG(module_name##_globals_id, zend_##module_name##_globals *, v)
```

Where `ZEND_TSMRG` is defined in `Zend/zend.h` to be `TSRMG`, or `TSRMG_STATIC` (if `ZEND_ENABLE_STATIC_TSRMLS_CACHE` is defined).  The difference between those two just seems to be whether the address of the cache is in a variable or fetched by calling a function.

PHP 7.0 and 7.2 seem to be the same.  What version of PHP are you using?

Reply:
Without this patch, swig generates this code:
#define SWIG_ErrorMsg() (example_globals.error_msg)
#define SWIG_ErrorCode() (example_globals.error_code)
Trying to compile this on Windows with thread-safe PHP will not work. Here is the output:
```
examplePHP7_wrap.c
C:\dev\work\swig-example\build\examplePHP7_wrap.c(952): warning C4311: 'type cast': pointer truncation from 'swig_module_info *' to 'long'
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2224: left of '.error_code' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): warning C4047: 'function': 'int' differs in levels of indirection from 'char [3]'
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): warning C4024: 'zend_error': different types for formal and actual parameter 1
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2224: left of '.error_msg' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2198: 'zend_error': too few arguments for call
C:\dev\work\swig-example\build\examplePHP7_wrap.c(987): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(987): error C2224: left of '.error_msg' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(988): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(988): error C2224: left of '.error_code' must have struct/union type
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~2\2017\COMMUN~1\VC\Tools\MSVC\1413~1.261\bin\Hostx64\x64\cl.exe' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
```
This is the code from zend_API.h from PHP source (latest git-version, 7.3.0-dev):
```
#ifdef ZTS

#define ZEND_DECLARE_MODULE_GLOBALS(module_name)							\
	ts_rsrc_id module_name##_globals_id;
#define ZEND_EXTERN_MODULE_GLOBALS(module_name)								\
	extern ts_rsrc_id module_name##_globals_id;
#define ZEND_INIT_MODULE_GLOBALS(module_name, globals_ctor, globals_dtor)	\
	ts_allocate_id(&module_name##_globals_id, sizeof(zend_##module_name##_globals), (ts_allocate_ctor) globals_ctor, (ts_allocate_dtor) globals_dtor);
#define ZEND_MODULE_GLOBALS_ACCESSOR(module_name, v) ZEND_TSRMG(module_name##_globals_id, zend_##module_name##_globals *, v)
#if ZEND_ENABLE_STATIC_TSRMLS_CACHE
#define ZEND_MODULE_GLOBALS_BULK(module_name) TSRMG_BULK_STATIC(module_name##_globals_id, zend_##module_name##_globals *)
#else
#define ZEND_MODULE_GLOBALS_BULK(module_name) TSRMG_BULK(module_name##_globals_id, zend_##module_name##_globals *)
#endif

#else

#define ZEND_DECLARE_MODULE_GLOBALS(module_name)							\
	zend_##module_name##_globals module_name##_globals;
#define ZEND_EXTERN_MODULE_GLOBALS(module_name)								\
	extern zend_##module_name##_globals module_name##_globals;
#define ZEND_INIT_MODULE_GLOBALS(module_name, globals_ctor, globals_dtor)	\
	globals_ctor(&module_name##_globals);
#define ZEND_MODULE_GLOBALS_ACCESSOR(module_name, v) (module_name##_globals.v)
#define ZEND_MODULE_GLOBALS_BULK(module_name) (&module_name##_globals)

#endif
```
Depending on ZTS set or not, it defines (with module-name example), example_globals or example_globals_id. So swig needs to do the same.
I'm not sure, why this is only a problem on Windows. Of course, it is still possible, that i'm doing something wrong but i spend really a lot of hours to find this and if i patch swig like this, i can compile the extension and load it into apache using php as module.
Building php without thread-safety is no problem.

If you (or anyone else want), I can provide you my batch-files to build php and the extension on windows. it is relatively easy.
And of course, i can test for php 7.0/7.1/7.2 to see if the situation is the same.

Reply:
With php 7.2 x64 thread-safe on Windows, it's the same.
Trying to compile without this patch produces:
```
[ 80%] Building C object CMakeFiles/php_example.dir/examplePHP7_wrap.c.obj
examplePHP7_wrap.c
C:\dev\work\swig-example\build\examplePHP7_wrap.c(952): warning C4311: 'type cast': pointer truncation from 'swig_module_info *' to 'long'
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2224: left of '.error_code' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): warning C4047: 'function': 'int' differs in levels of indirection from 'char [3]'
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): warning C4024: 'zend_error': different types for formal and actual parameter 1
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2224: left of '.error_msg' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(978): error C2198: 'zend_error': too few arguments for call
C:\dev\work\swig-example\build\examplePHP7_wrap.c(987): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(987): error C2224: left of '.error_msg' must have struct/union type
C:\dev\work\swig-example\build\examplePHP7_wrap.c(988): error C2065: 'example_globals': undeclared identifier
C:\dev\work\swig-example\build\examplePHP7_wrap.c(988): error C2224: left of '.error_code' must have struct/union type
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~2\2017\COMMUN~1\VC\Tools\MSVC\1413~1.261\bin\Hostx64\x64\cl.exe' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
```

Reply:
> Without this patch, swig generates this code:
> #define SWIG_ErrorMsg() (example_globals.error_msg)
> #define SWIG_ErrorCode() (example_globals.error_code)

Look at your patch - the code you are replacing doesn't generate the above at all, it generates this:

```
#define SWIG_ErrorMsg() ZEND_MODULE_GLOBALS_ACCESSOR(example, error_msg)
#define SWIG_ErrorCode() ZEND_MODULE_GLOBALS_ACCESSOR(example, error_code)
```

Looking at the history of this code, this was changed in 224bb9e0232a219b7cefd23472b592b39e277f46 - prior to that we did generate the code you show.

So this should already work with git master (and will in 4.0.0).  The proposed change doesn't achieve anything useful, so closing.

Reply:
Olly,

you are right. I didn't test with git master before changing it. Tested it right now and it works.
Thank you!